### ISSUE TITLE
How to evaluate the learned embedding space and skill prior?

### ISSUE THREAD

**Issue by yquantao** [clvrai/spirl#14]
Hi,

before we train the RL policy, how could we evaluate the skill embedding space **_Z_** and the learned skill prior? I know that the training results can be visualized via tensorboard, but do we have other metrics to check its performance or how could we make sure the skill prior really work?

**Comment by kpertsch** [comment]
Hi! 

Sorry for the slightly delayed reply! 

As a sanity check I usually perform "prior rollouts", ie I run the RL loop but without updating the high-level policy (the high-level policy by default gets initialized to the learned skill prior). Then by looking at (1) the rollout videos on wandb, (2) the average obtained reward or (3) the trajectory plots in the maze environment you can judge whether the combination of learned skills and skill prior leads to meaningful exploration. I found this to be usually much more informative of downstream task performance than any of the pre-training metrics like reconstruction loss or NLL of the learned skill prior.

The easiest way to run "prior rollouts" is by skipping the policy updates [here](https://github.com/clvrai/spirl/blob/581db4030989145c32bf0390cd9a1aec0f9cd0dd/spirl/rl/agents/ac_agent.py#L152) (e.g. by commenting out this line).

In the maze environment I would expect a successfully trained prior to produce rollout trajectories that widely explore the maze (like panel 2 in Fig. 5 of the paper). In the kitchen I would expect the prior rollout policy to generally perform meaningful robot movements in the videos and successfully perform 1-2 subtasks per rollout trajectory.

**Comment by yquantao** [comment]
Hi,

thanks for these helpful hints. I asked this question because when I trained skill embedding and skill prior with my own dataset(160 trajectories), the reconstructed outputs from the decoder are exactly the same. The first insight I could think is the KL divergence weight should be tuned, I tried with different KL weight values, but I still have this problem, do you have any other idea about this?

**Comment by yquantao** [comment]
I just figured out the actions in my dataset are not normalized, now the reconstructed values are reasonable, thank you.

**Comment by kpertsch** [comment]
Nice, glad you figured it out! Will close this issue for now, feel free to reopen it if you have more questions!
